### PR TITLE
[gql_websocket_link] enhancement for connection errors

### PR DESCRIPTION
**Pull request by dehypnosis**
Hi,

At First, Currently Subscription requests are losing message streams when having WebSocketChannelException.
And this problem is cannot be recovered automatically even if autoReconnect is set true, so the websocket connection recovered later.
It means when client have a connection problem, all message stream subscriptions for 'Subscription Operation' should be manually recreated.
Fixed this issue by not closing message subscription on WebSocketChannelException.

Secondly, Here added `StreamController<ConnectionError> _connectionErrorController` and public accessor `connectionErrorStream`.
A practical use case is that, when web socket connection failed due to server side decision or unexpected error, client can be notified about detail of error to take some action.
For example from my case, I need to refresh my client-side id token when it has been expired. I am using Firebase Authentication, and I can make Firebase detect token expiration on App startup. But when app became active from background after more than one hour, I should be detect token is expired or revoked by server..

I made a PR to deal with above case, or any other server-side connection errors.
Please take a consideration of this PR and thank you for reviewing it.

### CODE REVIEW

**Comment by knaeckeKami** [comment]
Thanks for your PR. 
It looks like the streamcontroller created is never disposed,  even it the link is disposed. could you fix that?

**Comment by agent3bood** [comment]
Thanks for taking time to describe the issue and providing a PR.

We are already forwarding the errors to the same stream here, and you can listen to it as well.
With your addition does that bring new functionality to the library?
```
onError: (Object error) {
        _messagesController.addError(error);
 });
```

File: links/gql_websocket_link/lib/src/link.dart [hunk 1]
```diff
@@ -405,6 +417,7 @@ class WebSocketLink extends Link {
     await _channel?.sink.close(websocket_status.normalClosure);
     _connectionStateController.add(ConnectionState.closed);
     await _connectionStateController.close();
+    await _connectionErrorController.close();
```
Comment (dehypnosis) [inline links/gql_websocket_link/lib/src/link.dart:420]: Added disposing statement here.

File: links/gql_websocket_link/lib/src/link.dart [hunk 2]
```diff
@@ -245,6 +251,8 @@ class WebSocketLink extends Link {
             ).catchError(_messagesController.addError);
           });
           _reConnectRequests.clear();
+        } else if (parsedMessage is ConnectionError) {
+          _connectionErrorController.add(parsedMessage);
```
Comment (dehypnosis) [inline links/gql_websocket_link/lib/src/link.dart:255]: @agent3bood Thank you for the review.

As you mentioned, `onError` handler is already forwarding stream exceptions.
But `ConnectionError` seems not considered as an exception. And channel's `listen` handler currently handles `ConnectionAck` but not handle `ConnectionError`. So I added public connection error stream for the use-case handling ConnectionError manually.
https://github.com/gql-dart/gql/blob/master/links/gql_websocket_link/lib/src/link.dart#L237

Additionally, as I described in "First issue", when `_messagesController` got an 'error', all the subscription operations streams are closed now. So they cannot be recovered even after reconnection success.
So I added 277-280 Lines.

**Comment by dehypnosis** [comment]
And due to CI check failure (formatting error), I pushed reformatted codes.
But I don't know how to re-run CI checks... Could you help me guys?

**Pull request closed without merging**